doc/developer/design/20240625_source_versioning__table_from_sources.md

```diff
@@ -0,0 +1,537 @@
+to users after they have created a source to allow them to create any downstream tables
+from a selection of the available upstream references of that source.
+
+We have two options available to us for exposing this information to users:
+
+1.  Add a new statement to fetch from the upstream all possible 'upstream references' available
+    for ingestion.
+2.  Create a new system table that periodically updates with the set of available upstream
+    references in each sources, and a new statement to trigger a refresh of this system table.
+
+The 2nd option is more complex to implement, but presents the information in a form that
+is more similar to other similar information in Materialize. However since the data can
+become out of date, it's important that the refresh statement is highlighted prominently
+in user-facing documentation and related guides.
+
+The proposal for the 2nd option (a new system table):
+
+Introduce a `mz_catalog.mz_available_source_references` with the following columns:
+
+-   `source_id`: The source (e.g. a Postgres, MySQL, Kafka source) this reference is available for.
+-   `reference`: The 'external reference' for this available object, such that using `reference`
+    in a `CREATE TABLE <table_name> FROM SOURCE <source_id> (REFERENCE <reference>)` would
+    construct a table ingesting this upstream reference. This is essentially the fully-qualified
+    upstream name of the reference.
+-   `columns`: If available, an array of of columns that this external reference would have if it
+    were ingested as a table in Materialize. This will be NULL for source types where an
+    external registry is needed to determine the upstream schema (e.g. Kafka).
+-   `name`: The object name of the reference. E.g. for a Postgres source this is the table name,
+    and for kafka is the topic name.
+-   `namespace`: Any namespace that this reference object exists in, if applicable. For Postgres
+    this would be `<database>.<schema>`, for MySQL this would be `<schema>`, and for Kafka this
+    would be NULL.
+-   `last_updated_at`: The timestamp of when this reference was fetched from the upstream system.
+
+Note that we do not expose the types of any of the upstream columns, since each source has
+different constructs for how types are represented, and since some source-types can't
+represent all upstream types (e.g. MySQL) without knowing which of them will be represented
+as TEXT using a TEXT COLUMNS option. Therefore the list of `columns` does not imply that each
+of those columns is one that Materialize can actually ingest.
+
+A new `REFRESH SOURCE <source_id or name>` statement will be introduced that can trigger
+a refresh of the rows in `mz_catalog.mz_available_source_references` that pertain to the
+requested source. Otherwise, this catalog table will only be populated at source creation
+time and upon a defined refresh interval in the source operator (perhaps every 60 seconds).
```

Comment:
It's going to be hard to update the catalog table from a source operator, no? I might just punt on the automatic refresh and rely on manual refreshes to start. 

An easy improvement to the purely manual approach that's less work than an automatic background refresh is to have a `CREATE TABLE ... FROM SOURCE` command do a refresh if the metadata is more than _x_ seconds stale.

Reply:
A bare `REFRESH SOURCE` might be confusing. Should we make it `REFRESH SOURCE REFERENCES`, so it also blend into the name of the system catalog table?

Reply:
> It's going to be hard to update the catalog table from a source operator, no? I might just punt on the automatic refresh and rely on manual refreshes to start.

Yep it will be -- this makes it simpler!

> An easy improvement to the purely manual approach that's less work than an automatic background refresh is to have a CREATE TABLE ... FROM SOURCE command do a refresh if the metadata is more than x seconds stale.

Happy to add this but I'm not sure what the point is? The user wants to know the latest set of tables so that they know which `CREATE TABLE .. FROM SOURCE` statements can be run, so updating it after the fact just saves time for the next table, but they would still need to call the `REFRESH SOURCE REFERENCE` command first anyways

Reply:
Oh, does `CREATE TABLE ... FROM SOURCE` not rely on the cached metadata in the source references table, but always fetch it fresh from the upstream system? Anyway, it seems like if we ever fetch the upstream metadata, we might as well take the opportunity to update the metadata in the `mz_source_references` table.

Reply:
> Oh, does CREATE TABLE ... FROM SOURCE not rely on the cached metadata in the source references table, but always fetch it fresh from the upstream system?

Yes, the top-level source doesn't store schemas or information for all tables and we want to make sure we have the latest schema for the new table, so we always fetch the latest from the upstream system. It works the same as the `ALTER SOURCE .. ADD SUBSOURCE` statement does.

Reply:
From live discussion: We will just expose a `REFRESH ..` command and not attempt to refresh this table automatically at any other times for now. Additionally we'll put this table in `mz_internal` to avoid user confusion while we learn about whether there are any difficulties with the potential staleness semantics